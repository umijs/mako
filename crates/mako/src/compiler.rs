use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;
use std::sync::{Arc, Mutex, RwLock};
use std::time::Instant;

use anyhow::{anyhow, Error, Result};
use colored::Colorize;
use regex::Regex;
use swc_core::common::sync::Lrc;
use swc_core::common::{Globals, SourceMap, DUMMY_SP};
use swc_core::ecma::ast::Ident;
use swc_node_comments::SwcComments;
use tracing::debug;

use crate::ast::comments::Comments;
use crate::ast::file::win_path;
use crate::config::{Config, Mode, ModuleIdStrategy, OutputMode};
use crate::generate::chunk_graph::ChunkGraph;
use crate::generate::optimize_chunk::OptimizeChunksInfo;
use crate::module_graph::ModuleGraph;
use crate::plugin::{Plugin, PluginDriver, PluginGenerateEndParams};
use crate::plugins;
use crate::resolve::{get_resolvers, Resolvers};
use crate::share::helpers::SWC_HELPERS;
use crate::stats::StatsInfo;
use crate::utils::id_helper::{assign_numeric_ids, compare_modules_by_incoming_edges};
use crate::utils::{thread_pool, ParseRegex};

pub struct Context {
    pub module_graph: RwLock<ModuleGraph>,
    pub chunk_graph: RwLock<ChunkGraph>,
    pub assets_info: Mutex<HashMap<String, String>>,
    pub modules_with_missing_deps: RwLock<Vec<String>>,
    pub config: Config,
    pub numeric_ids_map: RwLock<HashMap<String, usize>>,
    pub args: Args,
    pub root: PathBuf,
    pub meta: Meta,
    pub plugin_driver: PluginDriver,
    pub stats_info: StatsInfo,
    pub resolvers: Resolvers,
    pub static_cache: RwLock<MemoryChunkFileCache>,
    pub optimize_infos: Mutex<Option<Vec<OptimizeChunksInfo>>>,
}

#[derive(Default)]
pub struct MemoryChunkFileCache {
    content_map: HashMap<String, (Vec<u8>, u64)>,
    root: Option<PathBuf>,
}

impl MemoryChunkFileCache {
    pub fn new(root: Option<PathBuf>) -> Self {
        Self {
            content_map: HashMap::new(),
            root,
        }
    }

    pub fn write<T: AsRef<str>>(&mut self, path: T, content: Vec<u8>, hash: u64) -> Result<()> {
        let str = path.as_ref();

        if let Some((_, in_mem_hash)) = self.content_map.get(str) {
            if *in_mem_hash != hash {
                self.write_to_disk(str, &content)?;
            }
        } else {
            self.write_to_disk(str, &content)?;
        }
        self.content_map
            .insert(path.as_ref().to_string(), (content, hash));
        Ok(())
    }

    pub fn read<T: AsRef<str>>(&self, path: T) -> Option<Vec<u8>> {
        self.content_map
            .get(path.as_ref())
            .map(|(content, _)| content.clone())
    }

    fn write_to_disk<T: AsRef<str>>(&self, path: T, content: &[u8]) -> Result<()> {
        if let Some(root) = &self.root {
            let path = root.join(path.as_ref());
            fs::write(path, content)?;
        }
        Ok(())
    }
}

#[derive(Default)]
pub struct Args {
    pub watch: bool,
}

impl Context {
    pub fn write_static_content<T: AsRef<str>>(
        &self,
        path: T,
        content: Vec<u8>,
        hash: u64,
    ) -> Result<()> {
        let mut map = self.static_cache.write().unwrap();
        map.write(path, content, hash)
    }

    pub fn get_static_content<T: AsRef<str>>(&self, path: T) -> Option<Vec<u8>> {
        let map = self.static_cache.read().unwrap();
        map.read(path)
    }
}

impl Default for Context {
    fn default() -> Self {
        let mut numeric_ids_map = HashMap::new();
        SWC_HELPERS.iter().enumerate().for_each(|(i, item)| {
            numeric_ids_map.insert(item.to_string(), i);
        });
        let config: Config = Default::default();
        let resolvers = get_resolvers(&config);
        Self {
            config,
            args: Args { watch: false },
            root: PathBuf::from(""),
            module_graph: RwLock::new(ModuleGraph::new()),
            chunk_graph: RwLock::new(ChunkGraph::new()),
            assets_info: Mutex::new(HashMap::new()),
            modules_with_missing_deps: RwLock::new(Vec::new()),
            meta: Meta::new(),
            plugin_driver: Default::default(),
            stats_info: StatsInfo::new(),
            resolvers,
            optimize_infos: Mutex::new(None),
            static_cache: Default::default(),
            numeric_ids_map: RwLock::new(numeric_ids_map),
        }
    }
}

pub struct Meta {
    pub script: ScriptMeta,
    pub css: CssMeta,
}

impl Meta {
    pub fn new() -> Self {
        Self {
            script: ScriptMeta::new(),
            css: CssMeta::new(),
        }
    }
}

impl Default for Meta {
    fn default() -> Self {
        Self::new()
    }
}

pub struct ScriptMeta {
    pub cm: Lrc<SourceMap>,
    pub origin_comments: RwLock<Comments>,
    pub output_comments: RwLock<Comments>,
    pub globals: Globals,
    // These idents may be used in other places, such as transform_async_module
    pub module_ident: Ident,
    pub exports_ident: Ident,
    pub require_ident: Ident,
}

impl ScriptMeta {
    fn new() -> Self {
        Self {
            cm: Default::default(),
            origin_comments: Default::default(),
            output_comments: Default::default(),
            globals: Globals::default(),
            module_ident: build_ident("module"),
            exports_ident: build_ident("exports"),
            require_ident: build_ident("__mako_require__"),
        }
    }
}

fn build_ident(ident: &str) -> Ident {
    Ident {
        ctxt: Default::default(),
        span: DUMMY_SP,
        sym: ident.into(),
        optional: false,
    }
}

pub struct CssMeta {
    pub cm: Lrc<SourceMap>,
    pub globals: Globals,
    pub comments: SwcComments,
}

impl CssMeta {
    fn new() -> Self {
        Self {
            cm: Default::default(),
            globals: Globals::default(),
            comments: Default::default(),
        }
    }
}

impl Context {
    pub fn emit_assets(&self, origin_path: String, output_path: String) {
        let mut assets_info = self.assets_info.lock().unwrap();
        assets_info.insert(origin_path, output_path);
    }
}

pub struct Compiler {
    pub context: Arc<Context>,
}

impl Compiler {
    pub fn new(
        config: Config,
        root: PathBuf,
        args: Args,
        extra_plugins: Option<Vec<Arc<dyn Plugin>>>,
    ) -> Result<Self> {
        if !root.is_absolute() {
            return Err(anyhow!("root path must be absolute"));
        }

        let root = PathBuf::from(win_path(root.to_str().unwrap()));

        // why add plugins before builtin plugins?
        // because plugins like less-loader need to be added before assets plugin
        // TODO: support plugin orders
        let mut plugins: Vec<Arc<dyn Plugin>> = vec![];
        if let Some(extra_plugins) = extra_plugins {
            plugins.extend(extra_plugins);
        }
        let builtin_plugins: Vec<Arc<dyn Plugin>> = vec![
            // features
            Arc::new(plugins::manifest::ManifestPlugin {}),
            Arc::new(plugins::copy::CopyPlugin {}),
            Arc::new(plugins::import::ImportPlugin {}),
            // file types
            Arc::new(plugins::context_module::ContextModulePlugin {}),
            Arc::new(plugins::runtime::MakoRuntime {}),
            Arc::new(plugins::invalid_webpack_syntax::InvalidWebpackSyntaxPlugin {}),
            Arc::new(plugins::hmr_runtime::HMRRuntimePlugin {}),
            Arc::new(plugins::wasm_runtime::WasmRuntimePlugin {}),
            Arc::new(plugins::async_runtime::AsyncRuntimePlugin {}),
            Arc::new(plugins::emotion::EmotionPlugin {}),
            Arc::new(plugins::tree_shaking::FarmTreeShake {}),
            Arc::new(plugins::detect_circular_dependence::LoopDetector {}),
        ];
        plugins.extend(builtin_plugins);

        let mut config = config;

        if config.mode == Mode::Production && config.experimental.imports_checker {
            plugins.push(Arc::new(plugins::imports_checker::ImportsChecker {}));
        }

        if let Some(progress) = &config.progress {
            plugins.push(Arc::new(plugins::progress::ProgressPlugin::new(
                plugins::progress::ProgressPluginOptions {
                    prefix: "Mako".to_string(),
                    template:
                        "‚óè {prefix:.bold} {bar:25.green/white.dim} ({percent}%) {wide_msg:.dim}"
                            .to_string(),
                    progress_chars: progress.progress_chars.clone(),
                },
            )));
        }

        if let Some(duplicate_package_checker) = &config.check_duplicate_package {
            plugins.push(Arc::new(
                plugins::duplicate_package_checker::DuplicatePackageCheckerPlugin::new()
                    .show_help(duplicate_package_checker.show_help)
                    .emit_error(duplicate_package_checker.emit_error)
                    .verbose(duplicate_package_checker.verbose),
            ));
        }

        if config.experimental.require_context {
            plugins.push(Arc::new(plugins::require_context::RequireContextPlugin {}))
        }

        if config.output.mode == OutputMode::Bundless {
            plugins.insert(
                0,
                Arc::new(plugins::bundless_compiler::BundlessCompilerPlugin {}),
            );
        }

        if std::env::var("DEBUG_GRAPH").is_ok_and(|v| v == "true") {
            plugins.push(Arc::new(plugins::graphviz::Graphviz {}));
        }

        if args.watch && std::env::var("SSU").is_ok_and(|v| v == "true") {
            plugins.push(Arc::new(plugins::ssu::SUPlus::new()));
        }

        if args.watch && config.experimental.central_ensure {
            plugins.push(Arc::new(plugins::central_ensure::CentralChunkEnsure {}));
        }

        if let Some(minifish_config) = &config._minifish {
            let inject = if let Some(inject) = &minifish_config.inject {
                let mut map = HashMap::new();

                for (k, ii) in inject.iter() {
                    map.insert(
                        k.clone(),
                        plugins::minifish::Inject {
                            from: ii.from.clone(),
                            name: k.clone(),
                            named: ii.named.clone(),
                            namespace: ii.namespace,
                            exclude: ii.exclude.parse_into_regex()?,
                            include: ii.include.parse_into_regex()?,
                            prefer_require: ii.prefer_require.map_or(false, |v| v),
                        },
                    );
                }
                Some(map)
            } else {
                None
            };

            plugins.insert(
                0,
                Arc::new(plugins::minifish::MinifishPlugin {
                    mapping: minifish_config.mapping.clone(),
                    meta_path: minifish_config.meta_path.clone(),
                    inject,
                }),
            );
        }

        let ignores = config
            .ignores
            .iter()
            .map(|ignore| Regex::new(ignore).map_err(Error::new))
            .collect::<Result<Vec<Regex>>>()?;
        plugins.push(Arc::new(plugins::ignore::IgnorePlugin { ignores }));

        let plugin_driver = PluginDriver::new(plugins);

        plugin_driver.modify_config(&mut config, &root, &args)?;

        let resolvers = get_resolvers(&config);
        let mut numeric_ids_map = HashMap::new();
        SWC_HELPERS.iter().enumerate().for_each(|(i, item)| {
            numeric_ids_map.insert(item.to_string(), i);
        });
        Ok(Self {
            context: Arc::new(Context {
                static_cache: if config.write_to_disk {
                    RwLock::new(MemoryChunkFileCache::new(Some(config.output.path.clone())))
                } else {
                    Default::default()
                },
                config,
                args,
                root,
                module_graph: RwLock::new(ModuleGraph::new()),
                chunk_graph: RwLock::new(ChunkGraph::new()),
                assets_info: Mutex::new(HashMap::new()),
                modules_with_missing_deps: RwLock::new(Vec::new()),
                meta: Meta::new(),
                plugin_driver,
                numeric_ids_map: RwLock::new(numeric_ids_map),
                stats_info: StatsInfo::new(),
                resolvers,
                optimize_infos: Mutex::new(None),
            }),
        })
    }

    pub fn compile(&self) -> Result<()> {
        // ÂÖàÊ∏ÖÁ©∫ dist ÁõÆÂΩï
        if self.context.config.clean {
            self.clean_dist()?;
        }

        let t_compiler = Instant::now();
        let start_time = chrono::Local::now().timestamp_millis();
        let building_with_message = format!(
            "Building with {} for {}...",
            "mako".to_string().cyan(),
            self.context.config.mode
        )
        .green();
        println!("{}", building_with_message);
        {
            crate::mako_profile_scope!("Build Stage");
            let files = self
                .context
                .config
                .entry
                .values()
                .map(|entry| {
                    let mut entry = entry.to_string_lossy().to_string();
                    let is_browser = matches!(
                        self.context.config.platform,
                        crate::config::Platform::Browser
                    );
                    let watch = self.context.args.watch;
                    let hmr = self.context.config.hmr.is_some();
                    if is_browser && watch && hmr {
                        entry = format!("{}?hmr", entry);
                    }
                    crate::ast::file::File::new_entry(entry, self.context.clone())
                })
                .collect();
            self.context.plugin_driver.build_start(&self.context)?;

            self.build(files)?;

            debug!("start after build");

            self.context
                .plugin_driver
                .after_build(&self.context, self)?;
        }

        self.context.plugin_driver.before_generate(&self.context)?;

        if let ModuleIdStrategy::Numeric = self.context.config.module_id_strategy {
            let module_graph = self.context.module_graph.read().unwrap();
            assign_numeric_ids(
                module_graph.modules(),
                |a, b| compare_modules_by_incoming_edges(&module_graph, &a.id, &b.id),
                |module, id| {
                    let mut numeric_ids_map = self.context.numeric_ids_map.write().unwrap();
                    // reserved ten indexes for swc helper and others runtime module
                    numeric_ids_map.insert(module.id.id.clone(), id + 10);
                },
            )
        }

        let result = {
            crate::mako_profile_scope!("Generate Stage");
            // need to put all rayon parallel iterators run in the existed scope, or else rayon
            // will create a new thread pool for those parallel iterators
            thread_pool::scope(|_| self.generate())
        };
        let t_compiler_duration = t_compiler.elapsed();
        match result {
            Ok(mut stats) => {
                stats.start_time = start_time;
                stats.end_time = chrono::Local::now().timestamp_millis();
                println!(
                    "{}",
                    format!(
                        "‚úì Built in {}",
                        format!("{}ms", t_compiler_duration.as_millis()).bold()
                    )
                    .green()
                );
                if !self.context.args.watch {
                    println!("{}", "Complete!".bold());
                }
                let params = PluginGenerateEndParams {
                    is_first_compile: true,
                    time: t_compiler.elapsed().as_millis() as i64,
                    stats,
                };
                self.context
                    .plugin_driver
                    .generate_end(&params, &self.context)?;
                self.context.plugin_driver.write_bundle(&self.context)?;
                Ok(())
            }
            Err(e) => Err(e),
        }
    }

    pub fn full_hash(&self) -> u64 {
        crate::mako_profile_function!();
        let cg = self.context.chunk_graph.read().unwrap();
        let mg = self.context.module_graph.read().unwrap();
        cg.full_hash(&mg)
    }

    fn clean_dist(&self) -> Result<()> {
        // compiler ÂâçÊ∏ÖÈô§ distÔºåÂ¶ÇÊûúÂêéÁª≠ dev ÁéØÂ¢É‰∏çÂú® output_path ÈáåÔºåÈúÄË¶ÅÂÜçË°•‰∏ä dev ÁöÑÈÄªËæë
        let output_path = &self.context.config.output.path;
        if fs::metadata(output_path).is_ok() {
            fs::remove_dir_all(output_path)?;
        }
        Ok(())
    }
}
