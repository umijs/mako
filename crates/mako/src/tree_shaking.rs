use std::collections::HashMap;

use swc_ecma_visit::VisitMutWith;
use tracing::debug;

use crate::compiler::Compiler;
use crate::module::ModuleId;
use crate::tree_shaking_module::{ModuleSystem, TreeShakingModule};
use crate::unused_statement_marker::UnusedStatementMarker;

impl Compiler {
    pub fn tree_shaking(&self) {
        // 拓扑排序
        let (entry_modules, sorted_modules, cycle_modules) = self.make_toposort();
        debug!("entry_modules: {:?}", &entry_modules);
        debug!("cycle_modules: {:?}", &cycle_modules);

        // 入口模块设置为副作用模块
        self.markup_entry_modules_as_side_effects(entry_modules);

        // 循环依赖模块设置为副作用
        self.markup_cycle_modules_as_side_effects(cycle_modules);

        let (tree_shaking_module_ids, mut tree_shaking_module_map) =
            self.create_tree_shaking_module_map(sorted_modules);

        let tree_shaking_module_map = &mut tree_shaking_module_map;

        let mut modules_to_remove = vec![];
        for tree_shaking_module_id in tree_shaking_module_ids {
            let tree_shaking_module = tree_shaking_module_map
                .get_mut(&tree_shaking_module_id)
                .unwrap();

            debug!(
                "tree_shaking module: [{}]{}",
                tree_shaking_module.side_effects, &tree_shaking_module_id.id,
            );
            debug!(
                "    - used_exports: {:?}",
                &tree_shaking_module.used_exports
            );

            if matches!(tree_shaking_module.module_system, ModuleSystem::ESModule) {
                // 包含副作用的模块
                if tree_shaking_module.side_effects {
                    {
                        {
                            let imports = tree_shaking_module.imports();
                            let exports = tree_shaking_module.exports();

                            // 分析使用情况
                            for import in imports {
                                debug!("    - import: {:?}", &import);
                                self.analyze_import_statement(
                                    tree_shaking_module_map,
                                    &tree_shaking_module_id,
                                    import,
                                );
                            }

                            for export in exports {
                                debug!("    - export: {:?}", &export);
                                self.analyze_export_statement(
                                    tree_shaking_module_map,
                                    &tree_shaking_module_id,
                                    export,
                                );
                            }
                        };
                    }
                } else {
                    // 不包含副作用的模块，执行 tree shaking
                    let tree_shaking_module = tree_shaking_module_map
                        .get_mut(&tree_shaking_module_id)
                        .unwrap();

                    let mut module_graph = self.context.module_graph.write().unwrap();
                    let module = module_graph
                        .get_module_mut(&tree_shaking_module.id)
                        .unwrap();
                    let ast = module.info.as_mut().unwrap().ast.as_script_mut();

                    // 通过 tree_shaking_module 进行无用的标记
                    let mut comments = self.context.meta.script.output_comments.write().unwrap();
                    let mut marker = UnusedStatementMarker::new(tree_shaking_module, &mut comments);
                    ast.visit_mut_with(&mut marker);
                    drop(module_graph);
                    drop(comments);

                    // 当前模块没有被使用到的导出，删除当前模块
                    if tree_shaking_module.used_exports.is_empty() {
                        modules_to_remove.push(tree_shaking_module_id.clone());
                        continue;
                    }

                    {
                        let imports = tree_shaking_module.imports();
                        let exports = tree_shaking_module.exports();

                        // 分析使用情况
                        for import in imports {
                            debug!("    - import: {:?}", &import);
                            self.analyze_import_statement(
                                tree_shaking_module_map,
                                &tree_shaking_module_id,
                                import,
                            );
                        }

                        for export in exports {
                            debug!("    - export: {:?}", &export);
                            self.analyze_export_statement(
                                tree_shaking_module_map,
                                &tree_shaking_module_id,
                                export,
                            );
                        }
                    };
                }
            }
        }

        self.cleanup_no_used_export_module(modules_to_remove);
    }

    fn make_toposort(&self) -> (Vec<ModuleId>, Vec<ModuleId>, Vec<Vec<ModuleId>>) {
        let module_graph = self.context.module_graph.read().unwrap();
        let entry_modules = module_graph.get_entry_modules();
        let (sorted_modules, cycle_modules) = module_graph.toposort();
        (entry_modules, sorted_modules, cycle_modules)
    }

    fn markup_entry_modules_as_side_effects(&self, entry_modules: Vec<ModuleId>) {
        let mut module_graph = self.context.module_graph.write().unwrap();
        for entry_module_id in entry_modules {
            let module = module_graph.get_module_mut(&entry_module_id).unwrap();
            module.side_effects = true;
        }
    }

    fn markup_cycle_modules_as_side_effects(&self, cycle_modules: Vec<Vec<ModuleId>>) {
        let mut module_graph = self.context.module_graph.write().unwrap();
        for cycle_module in cycle_modules {
            for module_id in cycle_module {
                let module = module_graph.get_module_mut(&module_id).unwrap();
                module.side_effects = true;
            }
        }
    }

    #[allow(dead_code)]
    fn cleanup_no_used_export_module(&self, modules_to_remove: Vec<ModuleId>) {
        let mut module_graph = self.context.module_graph.write().unwrap();
        debug!("modules_to_remove: {:?}", &modules_to_remove);
        for module_id in modules_to_remove {
            module_graph.mark_missing_module(&module_id, &self.context);
        }
    }

    fn create_tree_shaking_module_map(
        &self,
        sorted_modules: Vec<crate::module::ModuleId>,
    ) -> (Vec<ModuleId>, HashMap<ModuleId, TreeShakingModule>) {
        let mut tree_shaking_module_ids = vec![];
        let mut tree_shaking_module_map = HashMap::new();
        for module_id in sorted_modules {
            let mut dependencies_of_module: Vec<ModuleId> = vec![];
            {
                let module_graph_r = self.context.module_graph.read().unwrap();
                dependencies_of_module.extend(
                    module_graph_r
                        .get_dependencies(&module_id)
                        .iter()
                        .map(|f| f.0.clone())
                        .collect::<Vec<_>>(),
                );
            }

            let mut module_graph = self.context.module_graph.write().unwrap();
            let module = module_graph.get_module_mut(&module_id).unwrap();

            // external 模块，设置为副作用模块
            if module.is_external() {
                module.side_effects = true;
            }

            if !module.get_module_type().is_script() {
                // 非 js 模块，以及他们的依赖，都设置为副作用模块
                for dep_module_id in dependencies_of_module {
                    let dep_module = module_graph.get_module_mut(&dep_module_id).unwrap();
                    dep_module.side_effects = true;
                }
            }

            let module = module_graph.get_module(&module_id).unwrap();
            let tree_shaking_module = TreeShakingModule::new(module);
            tree_shaking_module_ids.push(tree_shaking_module.id.clone());
            tree_shaking_module_map.insert(tree_shaking_module.id.clone(), tree_shaking_module);
        }
        (tree_shaking_module_ids, tree_shaking_module_map)
    }
}

#[cfg(test)]
mod tests {
    use crate::assert_display_snapshot;
    use crate::test_helper::{read_dist_file, setup_compiler};

    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking() {
        let compiler = setup_compiler("test/build/tree-shaking", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_reexport() {
        let compiler = setup_compiler("test/build/tree-shaking_reexport", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_named_reexport() {
        let compiler = setup_compiler("test/build/tree-shaking_named_reexport", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_export_namespace() {
        let compiler = setup_compiler("test/build/tree-shaking_export_namespace", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_named_export() {
        let compiler = setup_compiler("test/build/tree-shaking_named_export", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_fn() {
        let compiler = setup_compiler("test/build/tree-shaking_fn", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_side_effect() {
        let compiler = setup_compiler("test/build/tree-shaking_side_effect", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_class() {
        let compiler = setup_compiler("test/build/tree-shaking_class", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_exported() {
        let compiler = setup_compiler("test/build/tree-shaking_exported", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }
    #[tokio::test(flavor = "multi_thread")]
    async fn test_tree_shaking_export_default() {
        let compiler = setup_compiler("test/build/tree-shaking_export_default", false);
        compiler.compile();
        let content = read_dist_file(&compiler);
        assert_display_snapshot!(content);
    }
}
