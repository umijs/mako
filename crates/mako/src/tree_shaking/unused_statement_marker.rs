use std::collections::HashSet;

use mako_core::swc_ecma_ast;
use mako_core::swc_ecma_ast::{Decl, ExportDecl, ImportDecl, ModuleExportName, VarDeclarator};
use mako_core::swc_ecma_visit::{VisitMut, VisitMutWith, VisitWith};
use mako_core::tracing::debug;

use crate::comments::Comments;
use crate::tree_shaking::defined_ident_collector::DefinedIdentCollector;
use crate::tree_shaking::statement::{
    self, ExportSpecifier, ImportSpecifier, StatementId, StatementType,
};
use crate::tree_shaking::tree_shaking_analyze::strip_context;
use crate::tree_shaking::tree_shaking_module::{
    should_skip, TreeShakingModule, UsedIdent, UsedIdentHashMap,
};
/**
 * 针对没有使用到的 export、import 语句进行标记
 */
pub struct UnusedStatementMarker<'a, 'b> {
    tree_shaking_module: &'a TreeShakingModule,
    used_export_statement: Vec<(StatementId, HashSet<UsedIdent>)>,
    comments: &'b mut Comments,
}

impl<'a, 'b> UnusedStatementMarker<'a, 'b> {
    pub fn new(
        tree_shaking_module: &'a TreeShakingModule,
        used_export_statement: UsedIdentHashMap,
        comments: &'b mut Comments,
    ) -> Self {
        let used_export_statement = used_export_statement.into();
        Self {
            tree_shaking_module,
            used_export_statement,
            comments,
        }
    }
}

impl VisitMut for UnusedStatementMarker<'_, '_> {
    // 清理空模块，打上注释
    fn visit_mut_module(&mut self, module: &mut swc_ecma_ast::Module) {
        if should_skip(&self.tree_shaking_module.id.id) {
            debug!("skip module {:?}", &self.tree_shaking_module.id);
            return;
        }
        if self.tree_shaking_module.used_exports.is_empty() {
            self.comments.add_unused_module_comment(module.span.lo);
        } else {
            module.visit_mut_children_with(self);
        }
    }

    // 清理 export { } 这里面的变量
    fn visit_mut_export_specifiers(&mut self, specifiers: &mut Vec<swc_ecma_ast::ExportSpecifier>) {
        for (_, specifier) in specifiers.iter().enumerate() {
            match specifier {
                swc_ecma_ast::ExportSpecifier::Namespace(_) => {}
                swc_ecma_ast::ExportSpecifier::Default(_) => {}
                swc_ecma_ast::ExportSpecifier::Named(named_specifier) => {
                    let is_specifier_used = self.is_named_export_specifier_used(named_specifier);

                    if !is_specifier_used {
                        debug!("add unused comment to {:?}", &named_specifier);
                        self.comments.add_unused_comment(named_specifier.span.lo);
                    }
                }
            }
        }
    }

    fn visit_mut_export_default_expr(
        &mut self,
        default_expr: &mut swc_ecma_ast::ExportDefaultExpr,
    ) {
        if !self.tree_shaking_module.used_exports.contains(&"default") {
            debug!("add unused comment to default fn");
            self.comments.add_unused_comment(default_expr.span.lo)
        }
    }

    fn visit_mut_export_default_decl(
        &mut self,
        default_decl: &mut swc_ecma_ast::ExportDefaultDecl,
    ) {
        if !self.tree_shaking_module.used_exports.contains(&"default") {
            debug!("add unused comment to default fn");
            self.comments.add_unused_comment(default_decl.span.lo)
        }
    }

    fn visit_mut_export_decl(&mut self, export_decl: &mut ExportDecl) {
        match &mut export_decl.decl {
            Decl::Var(var_decl) => {
                for decl in &var_decl.decls {
                    let is_decl_used = self.is_var_decl_used(decl);

                    if !is_decl_used {
                        debug!("add unused comment to {:?}", &decl.name);
                        self.comments.add_unused_comment(decl.span.lo);
                    }
                }
            }
            Decl::Class(class_decl) => {
                let is_decl_used = self.is_fn_or_class_decl_ident_used(&class_decl.ident);

                if !is_decl_used {
                    debug!("add unused comment to {:?}", &class_decl.ident);
                    self.comments.add_unused_comment(class_decl.ident.span.lo);
                }
            }
            Decl::Fn(fn_decl) => {
                let is_decl_used = self.is_fn_or_class_decl_ident_used(&fn_decl.ident);

                if !is_decl_used {
                    debug!("add unused comment to {:?}", &fn_decl.ident);
                    self.comments.add_unused_comment(fn_decl.ident.span.lo)
                }
            }
            _ => (),
        }
    }

    fn visit_mut_import_decl(&mut self, import_decl: &mut ImportDecl) {
        if should_skip(&import_decl.src.value) {
            return;
        }
        // 清理没有用到的 specifier
        for specifier in &import_decl.specifiers {
            if let swc_ecma_ast::ImportSpecifier::Named(named_specifier) = specifier {
                let is_specifier_used =
                    &self.used_export_statement.iter().any(|(statement_id, ..)| {
                        let statement = self.tree_shaking_module.get_statement(statement_id);
                        if let StatementType::Import(import_statement) = statement {
                            is_same_import_specifier(import_statement, named_specifier)
                        } else {
                            false
                        }
                    });

                if !is_specifier_used {
                    debug!("add unused comment to {:?}", &named_specifier);
                    self.comments.add_unused_comment(named_specifier.span.lo)
                }
            }
        }
    }
}

impl<'a, 'b> UnusedStatementMarker<'a, 'b> {
    fn is_named_export_specifier_used(
        &mut self,
        named_specifier: &swc_ecma_ast::ExportNamedSpecifier,
    ) -> bool {
        self.used_export_statement
            .iter()
            .any(|(statement_id, used_idents)| {
                let statement = self.tree_shaking_module.get_statement(statement_id);
                if let StatementType::Export(export_statement) = statement {
                    let local = match &named_specifier.orig {
                        ModuleExportName::Ident(i) => i.clone(),
                        ModuleExportName::Str(_) => {
                            unreachable!("str as ident is not supported")
                        }
                    };
                    if !is_same_export_ident(export_statement, &local) {
                        return false;
                    }
                    // 可能在 export {} 中使用部分变量，需要单独把他们识别出来
                    let exported = if let Some(exported) = &named_specifier.exported {
                        exported
                    } else {
                        &named_specifier.orig
                    };
                    let local = match exported {
                        ModuleExportName::Ident(i) => i.clone(),
                        ModuleExportName::Str(_) => {
                            unreachable!("str as ident is not supported")
                        }
                    };

                    used_idents.contains(&UsedIdent::SwcIdent(local.to_string()))
                } else {
                    false
                }
            })
    }
}

impl<'a, 'b> UnusedStatementMarker<'a, 'b> {
    fn is_var_decl_used(&mut self, decl: &VarDeclarator) -> bool {
        let is_decl_used = &self.used_export_statement.iter().any(|(statement_id, ..)| {
            let statement = self.tree_shaking_module.get_statement(statement_id);
            if let StatementType::Export(export_statement) = statement {
                is_same_export_decl(export_statement, decl)
            } else {
                false
            }
        });
        *is_decl_used
    }

    fn is_fn_or_class_decl_ident_used(&mut self, ident: &dyn ToString) -> bool {
        let is_decl_used = &self.used_export_statement.iter().any(|(statement_id, ..)| {
            let statement = self.tree_shaking_module.get_statement(statement_id);
            if let StatementType::Export(export_statement) = statement {
                is_same_export_ident(export_statement, ident)
            } else {
                false
            }
        });
        *is_decl_used
    }
}

fn is_same_export_decl(
    export_statement: &statement::ExportStatement,
    decl: &dyn VisitWith<DefinedIdentCollector>,
) -> bool {
    let mut defined_ident_collector = DefinedIdentCollector::new();
    decl.visit_with(&mut defined_ident_collector);

    export_statement
        .info
        .specifiers
        .iter()
        .any(|export_specifier| match export_specifier {
            ExportSpecifier::Named { local, exported } => {
                if let Some(exported) = exported {
                    defined_ident_collector.defined_ident.contains(exported)
                } else {
                    defined_ident_collector.defined_ident.contains(local)
                }
            }
            _ => false,
        })
}

fn is_same_export_ident(
    export_statement: &statement::ExportStatement,
    ident: &dyn ToString,
) -> bool {
    export_statement
        .info
        .specifiers
        .iter()
        .any(|export_specifier| match export_specifier {
            ExportSpecifier::Named { local, .. } => ident.to_string() == *local,
            ExportSpecifier::Default => strip_context(&ident.to_string()) == "default",
            _ => false,
        })
}

fn is_same_import_specifier(
    import_statement: &crate::tree_shaking::statement::ImportStatement,
    named_specifier: &swc_ecma_ast::ImportNamedSpecifier,
) -> bool {
    import_statement
        .info
        .specifiers
        .iter()
        .any(|specifier| match specifier {
            ImportSpecifier::Named { local, .. } => named_specifier.local.to_string() == *local,
            _ => false,
        })
}
