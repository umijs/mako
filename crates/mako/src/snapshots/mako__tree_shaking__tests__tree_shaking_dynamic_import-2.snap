---
source: crates/mako/src/tree_shaking.rs
expression: content
---
globalThis.jsonpCallback([
    [
        './a.ts'
    ],
    {
        "./a.css": function(module, exports, require) {},
        "./a.ts": function(module, exports, require) {
            "use strict";
            Object.defineProperty(exports, "__esModule", {
                value: true
            });
            require("./a.css");
        }
    }
]);

//# sourceMappingURL=a_ts-async.js.map
